compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt

```diff
@@ -19,9 +19,45 @@ import org.jetbrains.kotlin.fir.declarations.impl.FirDefaultPropertyAccessor
 import org.jetbrains.kotlin.fir.types.FirImplicitTypeRef
 import org.jetbrains.kotlin.lexer.KtTokens
 
+internal fun isInsideExpectClass(containingClass: FirRegularClass, context: CheckerContext): Boolean =
+    isInsideSpecificClass(containingClass, context) { klass -> klass.isExpect }
+
+internal fun isInsideExternalClass(containingClass: FirRegularClass, context: CheckerContext): Boolean =
+    isInsideSpecificClass(containingClass, context) { klass -> klass.isExternal }
+
 // Note that the class that contains the currently visiting declaration will *not* be in the context's containing declarations *yet*.
-internal fun isInsideExpectClass(containingDeclaration: FirRegularClass, context: CheckerContext): Boolean =
-    containingDeclaration.isExpect || context.containingDeclarations.asReversed().any { it is FirRegularClass && it.isExpect }
+private inline fun isInsideSpecificClass(
+    containingClass: FirRegularClass,
+    context: CheckerContext,
+    specificStatus: (FirRegularClass) -> Boolean
+): Boolean =
+    specificStatus.invoke(containingClass) ||
+            context.containingDeclarations.asReversed().any { it is FirRegularClass && specificStatus.invoke(it) }
+
+internal fun FirMemberDeclaration.isEffectivelyExpect(
+    containingClass: FirRegularClass?,
+    modifierList: FirModifierList? = null,
+    context: CheckerContext,
+): Boolean {
+    val isExpect = this.isExpect || modifierList?.modifiers?.any { it.token == KtTokens.EXPECT_KEYWORD } == true
```

Comment:
Oh, I'm the one who spreads the wrong myth. cc @tgeng and @punzki just in case.

So, it all starts from some conflicting _modality_ modifiers, e.g., abstract final something. Then, found a similar case for _visibility_ as well. Then, I somehow generalize that and use it everywhere. For those `expect` and `external` (and I guess all other resolve modifiers), `status` is indeed good enough. The 5th commit cleaned up all the unnecessary checks I added.